web/web.go

```diff
@@ -0,0 +1,773 @@
+		return
+	}
+	// Filter typeMap so it contains just the types SrcSchema uses.
+	filteredTypeMap := make(map[string][]typeIssue)
+	for _, srcTable := range app.conv.SrcSchema {
+		for _, colDef := range srcTable.ColDefs {
+			if _, ok := filteredTypeMap[colDef.Type.Name]; ok {
+				continue
+			}
+			filteredTypeMap[colDef.Type.Name] = typeMap[colDef.Type.Name]
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(filteredTypeMap)
+}
+
+// setTypeMapGlobal allows to change spanner type globally.
+// It takes a map from source type to spanner type and updates
+// the spanner schema accordingly.
+func setTypeMapGlobal(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var typeMap map[string]string
+	err = json.Unmarshal(reqBody, &typeMap)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	// Redo source-to-Spanner type mapping using t (the mapping specified in the http request).
+	// We drive this process by iterating over the Spanner schema because we want to preserve all
+	// other customizations that have been performed via the UI (dropping columns, renaming columns
+	// etc). In particular, note that we can't just blindly redo schema conversion (using an appropriate
+	// version of 'toDDL' with the new type mapping).
+	for t, spSchema := range app.conv.SpSchema {
+		for col, _ := range spSchema.ColDefs {
+			sourceTable := app.conv.ToSource[t].Name
+			sourceCol := app.conv.ToSource[t].Cols[col]
+			srcCol := app.conv.SrcSchema[sourceTable].ColDefs[sourceCol]
+			if _, found := typeMap[srcCol.Type.Name]; found {
+				var ty ddl.Type
+				var issues []internal.SchemaIssue
+				switch app.driver {
+				case "mysql", "mysqldump":
+					ty, issues = toSpannerTypeMySQL(srcCol.Type.Name, typeMap[srcCol.Type.Name], srcCol.Type.Mods)
+				case "pg_dump", "postgres":
+					ty, issues = toSpannerTypePostgres(srcCol.Type.Name, typeMap[srcCol.Type.Name], srcCol.Type.Mods)
+				default:
+					http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+					return
+				}
+				if len(srcCol.Type.ArrayBounds) > 1 {
+					ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+					issues = append(issues, internal.MultiDimensionalArray)
+				}
+				if srcCol.Ignored.Default {
+					issues = append(issues, internal.DefaultValue)
+				}
+				if srcCol.Ignored.AutoIncrement {
+					issues = append(issues, internal.AutoIncrement)
+				}
+				if len(issues) > 0 {
+					app.conv.Issues[sourceTable][srcCol.Name] = issues
+				}
+				ty.IsArray = len(srcCol.Type.ArrayBounds) == 1
+				colDef := spSchema.ColDefs[col]
+				colDef.T = ty
+				spSchema.ColDefs[col] = colDef
+			}
+		}
+	}
+	updateSessionFile()
```

Comment:
Quick question: I assume we should call this function whenever we make changes to the session i.e. any mutations of state.

It is currently called here, and at end end of updateTableSchema and checkForInterleavedTables. Does this cover all of the relevant places?

Reply:
Yes nevin , these 3 are the only places where schema is changing as of now.